cryostatio/cryostat3
# feat(templates): reimplement custom event templates from mounted volume

# Welcome to Cryostat3! 👋
## Before contributing, make sure you have:

**To recreate commits with GPG signature** `git fetch upstream && git rebase --force --gpg-sign upstream/main`
_______________________________________________

Fixes: #431

## Description of the change:
Adds startup hook processing to the S3-based template storage service. It will scan the templates storage directory (`/opt/cryostat.d/templates.d` by default, same as earlier versions) at start and attempt to upload any files it recursively finds there to S3 storage as an event template. This produces a copy within S3 storage which can be used the same as any other template that the user manually uploads through the UI/API.

## Motivation for the change:
Part of 2.4 parity - this feature was enabled as a byproduct of old versions' direct filesystem storage of various resources, including custom event templates. The Operator used this to mount configmap files directly into the container's storage. With the move away from direct filesystem storage and instead using S3 for object storage this mechanism didn't work anymore, so this PR brings it back.

## How to manually test:
1. `cd cryostat3; mkdir templates; cp mytemplate.jfc templates` - to get a `mytemplate.jfc`, run the smoketest and download one of the existing event templates. Edit the resulting file and change the `label`, `description`, and `provider` attributes near the top, then move it to `cryostat3/mytemplate.jfc`.
2. Check out and build this PR
3. Run `./smoketest.bash -O`
4. Go to Events, select any connectable target, and ensure that your new `mytemplate.jfc` appears as a `CUSTOM` template in the table
5. Try to re-upload the same file again and it should be rejected, or any other template file with the same `label` attribute
6. ~~Try to delete the custom template and it should be rejected~~ in the latest version, deletions are accepted. They delete from S3 but leave the original copy on disk in place, so on next restart it would end up re-uploaded to S3.

In OpenShift,
1. `oc create configmap mytemplate --from-file=mytemplate.jfc`
2. Build and deploy Operator with `CORE_IMG=quay.io/andrewazores/cryostat:3.0.0-mounted-templates-1`
3. Create a Cryostat CR with the following:
```yaml
spec:
  eventTemplates:
    - configMapName: mytemplate
      filename: mytemplate.jfc
```
4. Wait for redeployment, then repeat steps 4-6 above

![Screenshot_2024-05-08_17-45-00](https://github.com/cryostatio/cryostat3/assets/3787464/f10035e1-3106-4fc4-b8de-454277ee2d5f)
